lib/matplotlib/tests/test_text.py

```diff
@@ -689,3 +689,21 @@ def test_fontproperties_kwarg_precedence():
     text2 = plt.ylabel("counts", size=40.0, fontproperties='Times New Roman')
     assert text1.get_size() == 40.0
     assert text2.get_size() == 40.0
+
+
+@check_figures_equal()
+def test_mutable_xy_copied_annotate(fig_test, fig_ref):
+    ax = fig_test.add_subplot()
+    ax.set_xlim(-5, 5)
+    ax.set_ylim(-3, 3)
+    xy_0 = np.array((-4, 1))
+    xy_f = np.array((-1, 1))
+    aprops = dict(arrowstyle='<->')
+    ax.annotate(text='', xy=xy_0, xytext=xy_f, arrowprops=aprops)
+    xy_0[1] = 3
+    ax = fig_ref.add_subplot()
+    ax.set_xlim(-5, 5)
+    ax.set_ylim(-3, 3)
+    xy_0 = np.array((-4, 1))
+    xy_f = np.array((-1, 1))
+    ax.annotate(text='', xy=xy_0, xytext=xy_f, arrowprops=aprops)
```

Comment:
Contradicting @tacaswell recommendation, I think it's sufficient to query the value:

~~~
xy = np.array([0.2, 0.2])
fig, ax = plt.subplots()
ann = ax.annotate('test', xy=xy, xytext=(0.8, 0.8), arrowprops=dict(arrowstyle='->'))
xy[0] = 0
assert ann.xy == (0,2, 0.2)
~~~

We can save the time and effort for generating a reference plot.

Reply:
Well, I have checked all text.py and the use of attribute `xy` of `_Annotationbase` (and its derivatives) is consistent with a tuple, and I have not seen something that could be broken with the tuple approach. `xy` **must** be indexable of at least 2 items, and if their type is something more sophisticated than plain numbers the conversion to tuple should work well. Which kind of data with units were you thinking of, @tacaswell? I am afraid I might be missing something here.

As for the test, at first I didn't like to directly query the internal representation of the xy placement, as it could change  later. Nevertheless `xy` is listed as an attribute for the `Annotation` derived class in the [docs](https://matplotlib.org/3.3.0/api/text_api.html#matplotlib.text.Annotation) and thus should be always accessible unless the API is changed. On the contrary, the test for `OffsetFrom` should be kept as it is now, since the internal representation of the point is not guaranteed by the documented API.

Unless @tacaswell has an objection, I'll revert to the tuple approach both in `_Annotationbase` and `OffsetFrom`, and adjust `_Annotationbase` test. I'd keep `OffsetFrom` test as it is now just to be on the safe side (or maybe too pedantic).

After this is resolved I'll take care of a similar issue in some classes of patches.py

Reply:
I assume "units" like `datetime64` or strings for categoricals.

Reply:
There is some data that we identify as "unitful" based on the types of the values in the data (datetime, datetime64, strings-as-catagorical), but there are other inputs that we identify as "uniful" based on the type or information carried on the container.  Due to the second case it is in general not safe to cast user input through to a fixed type.

Reply:
I think there is no problem with datetime, datetime64, strings-as-catagorical, as none is valid as an xy position. We're not talking about data being plotted here but about the annotation position.
I can check if it this cast  behaves well in the second case if someone gives me a hint to find an example of this unitful arrays. I guess that it could work if the conversion just translates an iterable to a tuple of objects of the same kind. The code in text.py just acceses `xy[0]` and `xy[1]`. Nothing fancier than that.

Reply:
I just checked an example using units (borrowed from [here](https://matplotlib.org/gallery/units/annotate_with_units.html#sphx-glr-gallery-units-annotate-with-units-py)) and the tuple keeps the type with units. Did you mean this kind of units, @tacaswell? If it is so, I can confirm the tuple approach works, and I'd go that way.

Reply:
Not following this part of the discussion. Is this still about whether we have to draw and compare images vs. just checkt the value on the artists?